### PR TITLE
bolt migrations

### PR DESCRIPTION
**Pull request by se3000**
Introduces new migration pattern for, to more easily upgrade past database versions.

### CODE REVIEW

File: store/migrations/migrate.go [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+	"sync"
+
+	"github.com/smartcontractkit/chainlink/logger"
+	"github.com/smartcontractkit/chainlink/store/migrations/migration1536521223"
+	"github.com/smartcontractkit/chainlink/store/migrations/migration1536696950"
+	"github.com/smartcontractkit/chainlink/store/orm"
+)
+
+func init() {
+	registerMigration(migration1536521223.Migration{})
```
Comment (rupurt) [inline store/migrations/migrate.go]: What's your thoughts on doing this at runtime? I was thinking you could read the list of migration directories from disk, get access to the migration in the package and then cast it to the interface through something like:

```golang
// loop through dirs on disk to get path
path := "github.com/smartcontractkit/chainlink/store/migrations/migration1536696950"
pkg, err := importer.Default().Import(path)
registerMigration (migration(pkg.Scope().Lookup("Migration")))
```

File: store/migrations/migration1536521223/migrate.go [hunk 2]
```diff
@@ -0,0 +1,235 @@
...
+)
+
+type Migration struct{}
+
+func (m Migration) Timestamp() string {
+	return "1536521223"
+}
+
+func (m Migration) Migrate(orm *orm.ORM) error {
+	orm.InitializeModel(&JobSpec{})
```
Comment (rupurt) [inline store/migrations/migration1536521223/migrate.go]: 👍 to redefining the types in the migration

File: store/migrations/migrate.go (hunk 1 above)
Comment (se3000) [reply store/migrations/migrate.go]: Reflection techniques are trickier in go than in other languages. An illustration of this is the fact that we don't have any migration directories to iterate over after compilation when we're running from our single go binary.

We tried having the `init` in the individual migration (`migrations/migration123`) self register, but if they're never imported in a top level pkg (`migrations`, or `store`) they are culled out and never get `init`d. The most common pattern we found was the explicit registration in an `init`. Let's chat more offline.

Comment (rupurt) [reply store/migrations/migrate.go]: Ah, good point about the dirs not being available after compilation. If it's too difficult or requires more research I'm ok with this approach.

File: store/migrations/migration0/migrate.go [hunk 3]
```diff
@@ -0,0 +1,294 @@
...
+type TaskType string
+
+type TaskSpec struct {
+	Type          TaskType    `json:"type" storm:"index"`
+	Confirmations uint64      `json:"confirmations"`
+	Params        models.JSON `json:"-"`
+}
+
+// UnmarshalJSON parses the given input and updates the TaskSpec.
+func (t *TaskSpec) UnmarshalJSON(input []byte) error {
```
Comment (j16r) [inline store/migrations/migration0/migrate.go]: Have we got a database which would trigger this code? I'm not sure it's actually doing anything.

File: store/migrations/migration0/migrate.go [hunk 4]
```diff
@@ -0,0 +1,294 @@
...
+func (w *WebURL) String() string {
+	if w.URL == nil {
+		return ""
+	}
+	return w.URL.String()
+}
+
+type Link big.Int
+
+func (l *Link) ToBig() *big.Int {
```
Comment (j16r) [inline store/migrations/migration0/migrate.go]: Is `MarshalText` used by boltdb? I think there is a bunch of dead code here.

**Pull request merged**